### DISCUSSION TITLE
Rule function return undefined value for the input param.

### DISCUSSION THREAD

**Discussion by benjammartin** [omar-dulaimi/trpc-shield#15] [General]
# Bug report

- [x] I have checked other issues to make sure this is not a duplicate.

## Describe the bug

Rule function return undefined for the input params. I'm using version `"^0.4.0"`.

```
const isRuleName = rule<Context>()(
  async (ctx, type, path, input, rawInput, options) => {
     console.log(input) // undefined
     return true
  }
);
```

## What I have 

```
const permissions = shield<Context>({
  project: {
    mutation: {
      delete: isRuleName,
    },
  }
});

export const permissionsMiddleware = middleware(permissions);
export const protectedProcedure = t.procedure.use(permissionsMiddleware);

const appRouter = router({
project: router({
    delete: protectedProcedure
      .input(z.object({ id: z.string(), owner: z.string() }))
      .mutation(async ({ ctx, input }) => {
        return await prisma.project.delete({
          where: {
            id: input.id,
          },
        });
      }),
   })

```

## Additional context

RawInput returns the content of the input.

**Comment by omar-dulaimi** [comment]
The value should come from tRPC itself. I'm not sure why it returns undefined here. Can you confirm that it actually return a value in other middlewares?

**Comment by benjammartin** [comment]
My bad! I didn't try with other middleware. This is also the case with others. I don't know why.

**Comment by omar-dulaimi** [comment]
Closing this, with the conclusion your own follow-up pointed at.

You confirmed that `input` is also undefined in other middlewares, not just in a shield rule. That places it in tRPC's middleware pipeline rather than in this library: shield does not compute `input`, it passes through whatever tRPC hands the middleware.

The relevant detail, which was not obvious in 2023: in a tRPC middleware, `input` is only populated **after** the input parser has run. A middleware attached before the `.input(...)` parser in the chain sees `undefined` there, which matches what you saw across several middlewares. `rawInput` is the unparsed value and is available either way, which is why shield passes both to your rule.

What has changed since, in case it matters for you now: shield tracks tRPC v11 as of 3.0.0, and the middleware reads `getRawInput()` rather than the old `rawInput` property, which v11 replaced.

If you still see `undefined` in a rule on a current version where you expect a parsed input, please open an issue with the procedure definition and the middleware order; that would be a real bug rather than pipeline ordering, and I would want it.
